### PR TITLE
Fix the vulnerability found in issue #3861

### PR DESCRIPTION
**Pull request by Lexterl33t**
# Fix the vulnerability found in issue #3861

## Description  
This PR addresses the **non-compliance issue** with the `PUSHN` opcodes in **EthereumJS**, as described in [issue #3861](https://github.com/ethereumjs/ethereumjs-monorepo/issues/3861).  

Previously, when a `PUSH` instruction had **fewer bytes available** than required, the missing bytes were **not properly zero-padded** to align with the **Ethereum Yellow Paper** specification.  

This led to inconsistencies between **EthereumJS** and other **EVM implementations** like **REVM** and **SputnikVM**, which correctly apply **right-aligned zero-padding**.

---

## Fix Implemented
- Ensured that `PUSHN` opcodes always **zero-pad missing bytes**, right-aligning the values as per the **Yellow Paper**.  
- Added a **test case** to validate that the stack correctly **returns the padded value**.

### CODE REVIEW

File: packages/evm/test/stack.spec.ts [hunk 1]
```diff
@@ -148,4 +148,33 @@ describe('Stack', () => {
...
+    await evm.stateManager.putCode(addr, hexToBytes(code))
+    await evm.stateManager.putAccount(caller, new Account(BigInt(0), BigInt(0x11)))
+ 
+    try {
+      const res = await evm.runCall({
+        value: BigInt(0),
+        data: hexToBytes(code),
+      });    
+      const executionStack = res.execResult.runState?.stack
+      console.log(executionStack)
```
Comment (scorbajio) [inline packages/evm/test/stack.spec.ts:174]:
```suggestion
```

File: packages/evm/src/opcodes/functions.ts [hunk 2]
```diff
@@ -935,12 +935,15 @@ export const handlers: Map<number, OpHandler> = new Map([
...
-        runState.stack.push(loaded)
-      }
+        const slice = new Uint8Array(numToPush); 
+        const availableBytes = Math.min(numToPush, runState.code.length - runState.programCounter);
+    
+        slice.set(runState.code.subarray(runState.programCounter, runState.programCounter + availableBytes));
+    
+        const loaded = bytesToBigInt(slice);
+        runState.programCounter += numToPush;
+        runState.stack.push(loaded);
```
Comment (jochem-brouwer) [inline packages/evm/src/opcodes/functions.ts:945]: These are some big changes to the PUSH opcode, which is ran a lot. We will need to keep performance in mind here, so I'll do some tests to see that the impact of this change is minimal.

File: packages/evm/test/stack.spec.ts [hunk 3]
```diff
@@ -148,4 +148,33 @@ describe('Stack', () => {
...
   })
+
+  it('stack should return the padded value', async () => {
+    const caller = new Address(hexToBytes('0x00000000000000000000000000000000000000ee'))
+    const addr = new Address(hexToBytes('0x00000000000000000000000000000000000000ff'))
+    const evm = await createEVM()
+    const account = createAccount(BigInt(0), BigInt(0))
+    const code = '0x6801'
+    const expectedReturnValue = new Stack(1024)
+    expectedReturnValue.push(BigInt(18446744073709551616))
```
Comment (jochem-brouwer) [inline packages/evm/test/stack.spec.ts:159]: This number seems magic, it would be helpful to have this constant being "calculated", for `bytesToBigInt(0x0100000000000000)` or an explanation why this should be this number (it is `2^64` ) would be helpful.

File: packages/evm/test/stack.spec.ts [hunk 4]
```diff
@@ -148,4 +148,33 @@ describe('Stack', () => {
...
+    /*
+      code:             
+          PUSH9 0x01
```
Comment (jochem-brouwer) [inline packages/evm/test/stack.spec.ts:162]: This code comment should be moved above the `code` param.

File: packages/evm/test/stack.spec.ts [hunk 5]
```diff
@@ -148,4 +148,33 @@ describe('Stack', () => {
...
+    */
+    await evm.stateManager.putAccount(addr, account)
+    await evm.stateManager.putCode(addr, hexToBytes(code))
+    await evm.stateManager.putAccount(caller, new Account(BigInt(0), BigInt(0x11)))
```
Comment (jochem-brouwer) [inline packages/evm/test/stack.spec.ts:166]: Why is the balance 11?

File: packages/evm/test/stack.spec.ts [hunk 6]
```diff
@@ -148,4 +148,33 @@ describe('Stack', () => {
...
+ 
+    try {
+      const res = await evm.runCall({
+        value: BigInt(0),
```
Comment (jochem-brouwer) [inline packages/evm/test/stack.spec.ts:170]: The value does not need to be explicitly set here, and should be removed.

File: packages/evm/test/stack.spec.ts [hunk 7]
```diff
@@ -148,4 +148,33 @@ describe('Stack', () => {
...
+        data: hexToBytes(code),
+      });    
+      const executionStack = res.execResult.runState?.stack
+      console.log(executionStack)
+      assert.deepEqual(executionStack, expectedReturnValue)
+    } catch (e: any) {
+      assert.fail(e.message)
```
Comment (jochem-brouwer) [inline packages/evm/test/stack.spec.ts:177]: The try/catch block can be eliminated, since the test will throw if an unhandled error is thrown, or if the `executionStack` does not deep equal `expectedReturnValue`

**Review by jochem-brouwer — COMMENTED** [review]
Some initial comments! :smile: :+1:

File: packages/evm/test/stack.spec.ts (hunk 3 above)
Comment (Lexterl33t) [reply packages/evm/test/stack.spec.ts]: Done ✅

File: packages/evm/test/stack.spec.ts (hunk 4 above)
Comment (Lexterl33t) [reply packages/evm/test/stack.spec.ts:156]: Done ✅

File: packages/evm/test/stack.spec.ts (hunk 5 above)
Comment (Lexterl33t) [reply packages/evm/test/stack.spec.ts]: Done ✅

File: packages/evm/test/stack.spec.ts (hunk 6 above)
Comment (Lexterl33t) [reply packages/evm/test/stack.spec.ts]: Done ✅

File: packages/evm/test/stack.spec.ts (hunk 7 above)
Comment (Lexterl33t) [reply packages/evm/test/stack.spec.ts]: Done ✅

**Review by jochem-brouwer — APPROVED** [review]

**Comment by jochem-brouwer** [comment]
LGTM, thanks, will leave this open for one more review. Have changed the code to run minimal extra logic per PUSH (only a length check) and in the edge case of the PUSH being at the end of the code, then `setLengthRight` will be called on the bytes. Have also expanded the tests to test all PUSH opcodes PUSH1-PUSH32 for this case.

**Review by jochem-brouwer — APPROVED** [review]

**Review by acolytec3 — APPROVED** [review]

**Review by acolytec3 — COMMENTED** [review]
LGTM
